---
layout: blog
authors: [ivagljiva, FlorianTrigodet]
title: Metabin refinement and population genetics using human tongue metagenomes
excerpt: "A multi-purpose tutorial based on human oral microbiome data"
date: 2025-03-05
tags: [metagenomics, binning, population genetics, hands-on, beginner]
comments: true
---

Real world datasets are complicated, and it can be difficult to accurately capture this in tutorials due to time and space constraints. The infamous [Infant Gut Tutorial](https://merenlab.org/tutorials/infant-gut/) covers the basics of both binning and population genetics on an extremely simple dataset, which is great for learning but the process and results are not necessarily representative of what one would actually end up doing/seeing in a large-scale 'omics study. Take binning, for example -- if you have dozens of large co-assemblies in your dataset, you won't have the time to do manual binning. And microbial communities containing hundreds of populations rarely show perfectly clean, easy-to-bin patterns of sequence composition and differential coverage. As for population genetics, our existing tutorial shows you how to do the thing, but leaves out some context for the question and important parameter decisions along the way.

With that in mind, we created this tutorial to show a more realistic example of how to work on both metagenomic binning and population genetics in anvi'o, using real-world data from the human oral microbiome as generated in the paper ["Functional and genetic markers of niche partitioning among enigmatic members of the human oral microbiome"](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02195-w) by Shaiber et al (2020). Our goal is to show you the following:

- How to run a read recruitment workflow from many samples to a co-assembly, to generate differential coverage data for binning as well as variant data for population genetics
- How to manually refine a 'metabin' that was automatically generated by Concoct and encompasses several microbial populations into a few smaller yet high-quality bins
- How to set some key parameters in a population genetics analysis that answers a (somewhat) realistic research question

Of course, we are still constrained a little bit by the time and computational resources allowed by a typical tutorial session. We will be working with a smallish co-assembly that is entirely visualizable in the anvi'o interactive interface (unlike many real-world datasets). And it is unlikely that we will actually run the read recruitment workflow in a tutorial setting (but we will give you all the tools to do so in case you want to try it at home).

Without further ado, let's begin :)

## Our dataset and questions

We are taking advantage of the [publicly-available data](https://merenlab.org/data/#niche-partitioning-in-the-human-oral-cavity) from Shaiber et al 2020. This study described a time-series of tongue and plague metagenomes taken from several human individuals, most of which were paired in male-female couples. This sampling strategy is particularly effective for metagenomic binning for the following reasons. First, having multiple samples from a single individual enabled the authors to do co-assembly, which effectively increases the coverage and likelihood of catching low-abundance populations (at the cost of increasing the overall complexity, which can break assembly algorithms). Second, we know that similar populations likely occur across different individuals and it therefore makes sense to map multiple individuals to a given co-assembly to leverage more differential coverage signals -- this was not the strategy used in the Shaiber et al 2020 paper, but it will work quite well for our purposes today.

Here are the questions we want to answer using this dataset:

1. What microbial populations can we recover high-quality metagenome-assembled genomes (MAGs) for by refining 'metabins'?
2. Do the individuals in a couple share similar populations in their oral microbiomes? To put it another way -- can the population variants specific to a given individual help us distinguish between the couples?

In order to answer these questions at a smaller scale, we will be using a subset of the data from the 2020 study - a co-assembly of 5 tongue samples from the time-series of a single individual (the sample named `T-B-M` in the original paper), plus 20 tongue metagenomes to recruit reads from. These include the 5 samples used to make the co-assembly and 5 each from three other individuals: `T-A-F`, `T-A-M`, and `T-B-F`.

For the purposes of this tutorial, we will use the co-assembly that was already generated by the study authors. We will start our binning journey from the results of the automatic binning tool [CONCOCT](https://github.com/BinPro/CONCOCT) ([Alneberg et al 2014](https://www.nature.com/articles/nmeth.3103)) that was already run on the co-assembly. 

In fact, CONCOCT was run twice -- once by specifying `-c 10` to force the tool to create exactly 10 'megabins' (each containing multiple microbial populations), and once without enforcing a number of bins (in which case the tool would try to group sequences from exactly one population in each bin). The 'metabin' strategy is extremely useful for partitioning your large metagenomic datasets into manageable chunks that can be manually refined, and is described elsewhere in [a blog post by Tom and Meren](https://anvio.org/blog/constrained-binning/).

## The mapping workflow

If all we wanted to do was bin refinement, a read recruitment step wouldn't really be necessary here. After all, the automatic binning results were already generated using existing read recruitment results where the 5 time-series samples were mapped against their co-assembly. So we could have just taken the existing, publicly-available contigs and profile databases from [this Figshare](https://figshare.com/articles/dataset/Anvi_o_profiles_per_individual/12217802) and done our bin refinement on that. 

But we want more than that. To answer question (2), we also want to see how the tongue-associated populations from the individual `T-B-M` differ from the tongue-associated populations in the three other individuals. And if we are going to map additional samples to the co-assembly anyway, might as well use them to help guide our bin refinement process.

